Mathlib/Algebra/Group/AddChar.lean

```diff
@@ -31,170 +30,236 @@ see `Mathlib.NumberTheory.LegendreSymbol.AddCharacter`.
+  DFunLike.ext f g h
 
-end DerivedInstances
+@[simp] lemma coe_mk (f : A → M)
+    (map_zero_one' : f 0 = 1) (map_add_mul' : ∀ a b : A, f (a + b) = f a * f b) :
+    AddChar.mk f map_zero_one' map_add_mul' = f := by
+  rfl
 
-section CoeToFun
+/-- An additive character maps `0` to `1`. -/
+@[simp] lemma map_zero_one (ψ : AddChar A M) : ψ 0 = 1 := ψ.map_zero_one'
+#align add_char.map_zero_one AddChar.map_zero_one
 
-variable {A : Type u} [AddMonoid A] {M : Type v} [CommMonoid M]
+/-- An additive character maps sums to products. -/
+@[simp] lemma map_add_mul (ψ : AddChar A M) (x y : A) : ψ (x + y) = ψ x * ψ y := ψ.map_add_mul' x y
+#align add_char.map_add_mul AddChar.map_add_mul
 
 /-- Interpret an additive character as a monoid homomorphism. -/
-def toMonoidHom : AddChar A M → Multiplicative A →* M :=
-  id
+def toMonoidHom (φ : AddChar A M) : Multiplicative A →* M where
+  toFun := φ.toFun
+  map_one' := φ.map_zero_one'
+  map_mul' := φ.map_add_mul'
 #align add_char.to_monoid_hom AddChar.toMonoidHom
 
-open Multiplicative
-
-/-- Define coercion to a function so that it includes the move from `A` to `Multiplicative A`.
-After we have proved the API lemmas below, we don't need to worry about writing `ofAdd a`
-when we want to apply an additive character. -/
-instance instFunLike : FunLike (AddChar A M) A M :=
-  inferInstanceAs (FunLike (Multiplicative A →* M) A M)
-#noalign add_char.has_coe_to_fun
+-- this instance was a bad idea and conflicted with `instFunLike` above
+#noalign add_char.monoid_hom_class
 
-theorem coe_to_fun_apply (ψ : AddChar A M) (a : A) : ψ a = ψ.toMonoidHom (ofAdd a) :=
+@[simp] lemma toMonoidHom_apply (ψ : AddChar A M) (a : Multiplicative A) : ψ.toMonoidHom a = ψ a :=
   rfl
-#align add_char.coe_to_fun_apply AddChar.coe_to_fun_apply
+#align add_char.coe_to_fun_apply AddChar.toMonoidHom_apply
 
--- Porting note: added
-theorem mul_apply (ψ φ : AddChar A M) (a : A) : (ψ * φ) a = ψ a * φ a :=
-  rfl
+/-- An additive character maps multiples by natural numbers to powers. -/
+@[simp] lemma map_nsmul_pow (ψ : AddChar A M) (n : ℕ) (x : A) : ψ (n • x) = ψ x ^ n :=
+  ψ.toMonoidHom.map_pow x n
+#align add_char.map_nsmul_pow AddChar.map_nsmul_pow
+
+variable (A M) in
+/-- Additive characters `A → M` are the same thing as monoid homomorphisms from `Multiplicative A`
+to `M`. -/
+def toMonoidHomEquiv : AddChar A M ≃ (Multiplicative A →* M) where
+  toFun φ := φ.toMonoidHom
+  invFun f :=
+  { toFun := f.toFun
+    map_zero_one' := f.map_one'
+    map_add_mul' := f.map_mul' }
+  left_inv _ := rfl
+  right_inv _ := rfl
+
+/-- Interpret an additive character as a monoid homomorphism. -/
+def toAddMonoidHom (φ : AddChar A M) : A →+ Additive M where
+  toFun := φ.toFun
+  map_zero' := φ.map_zero_one'
+  map_add' := φ.map_add_mul'
+
+variable (A M) in
+/-- Additive characters `A → M` are the same thing as additive homomorphisms from `A` to
+`Additive M`. -/
+def toAddMonoidHomEquiv : AddChar A M ≃ (A →+ Additive M) where
```

Comment:
Isn't it a bit asymmetric that for the map to ` Multiplicative A -> M`  we only have the equivalence and for the other one both? Perhaps it is enough to pack `toAddMonoidHom` in the equivalence and use that when needed.

Reply:
Fair enough, we can have `toAddMonoidHom` as a separate def, we may as well maintain symmetry.

Reply:
... I went to add this and realised it was already there? We have all four of `toAddMonoidHom`, `toAddMonoidHomEquiv`, `toMonoidHom`, and `toMonoidHomEquiv`. 

The one thing that is asymmetric is `toMonoidHomMulEquiv` (upgrading the `Equiv` to a `MulEquiv`) but that is because the choice to write the group law on `AddChar`'s as multiplication breaks the symmetry. I suppose we could have `toAddMonoidHomAddEquiv : Additive (AddChar A M) →+ (A →+ Additive M)` but I'm not sure how likely this is to be useful.

Reply:
I agree, probably not too useful.

Reply:
I added it anyway, for the following reason: it will make life easier if Yaël carries through his plan to rewrite `AddChar A M` as an additive rather than multiplicative monoid.